File: roman-arabic/functional-java/src/main/java/org/kata/ArabicNumber.java
package org.kata;

public final class ArabicNumber {

    private final int value;

    private ArabicNumber(int value) {
        this.value = value;
    }

    public static ArabicNumber arabicNumberOf(int value) {
        return new ArabicNumber(value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;

        if (o == null) {
            return false;
        }

        if (ArabicNumber.class == o.getClass()) {
            return value == ((ArabicNumber) o).value;
        }

        return Integer.class == o.getClass()
                && value == (Integer) o;

    }

    @Override
    public int hashCode() {
        return value;
    }
}

File: roman-arabic/scala/src/main/scala/org/kata/ArabicNumber.scala
package org.kata

case class ArabicNumber(private val value: Int) {

  def +(number: ArabicNumber) =
    ArabicNumber(value + number.value)

  def -(number: ArabicNumber) =
    ArabicNumber(value - number.value)

  def <(number: ArabicNumber) =
    value < number.value

  def canEqual(other: Any): Boolean =
    other.isInstanceOf[ArabicNumber] || other.isInstanceOf[Integer]

  override def equals(other: Any): Boolean = other match {
    case that: ArabicNumber =>
      (this canEqual that) &&
        value == that.value
    case that: Integer =>
      (this canEqual that) &&
        value == that
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(value)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
